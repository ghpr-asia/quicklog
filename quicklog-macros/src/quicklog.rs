use std::str::FromStr;

use proc_macro2::TokenStream as TokenStream2;
use quote::{quote, ToTokens};

pub(crate) enum ParseError {
    VariantNotFound,
}

// TODO: Duplicate of quicklog's Level, re-defined here to avoid including
// quicklog as a dependency
// Alternative is to break out a subset of quicklog into a separate
// `quicklog-core` crate or something similar, but that can be done in a future
// more general refactor.
#[derive(Copy, Clone, PartialOrd, PartialEq, Eq)]
pub(crate) enum Level {
    Trace = 0,
    Debug = 1,
    Info = 2,
    Warn = 3,
    Error = 4,
    Event = 5,
}

impl FromStr for Level {
    type Err = ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "trc" | "trace" | "0" => Ok(Level::Trace),
            "dbg" | "debug" | "1" => Ok(Level::Debug),
            "inf" | "info" | "2" => Ok(Level::Info),
            "wrn" | "warn" | "3" => Ok(Level::Warn),
            "err" | "error" | "4" => Ok(Level::Error),
            "evt" | "event" | "5" => Ok(Level::Event),
            _ => Err(ParseError::VariantNotFound),
        }
    }
}

impl ToTokens for Level {
    fn to_tokens(&self, tokens: &mut TokenStream2) {
        let tok = match self {
            Self::Trace => quote! { quicklog::level::Level::Trace },
            Self::Debug => quote! { quicklog::level::Level::Debug },
            Self::Info => quote! { quicklog::level::Level::Info },
            Self::Warn => quote! { quicklog::level::Level::Warn },
            Self::Error => quote! { quicklog::level::Level::Error },
            Self::Event => quote! { quicklog::level::Level::Event },
        };

        tokens.extend(tok);
    }
}
