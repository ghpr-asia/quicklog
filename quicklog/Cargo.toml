[package]
name = "quicklog"
version = "0.2.2"
# passing all test cases through trybuild
autotests = false
edition = "2021"

[lib]
name = "quicklog"
path = "src/lib.rs"

[[test]]
name = "ui"
path = "tests/ui.rs"

[[test]]
name = "derive"
path = "tests/derive/derive.rs"

[dependencies]
lazy_format = "2.0.0"
quicklog-clock = { path = "../quicklog-clock", version = "0.1.2", registry = "ghp-cargo-local" }
quicklog-flush = { path = "../quicklog-flush", version = "0.1.2", registry = "ghp-cargo-local" }
quicklog-macros = { path = "../quicklog-macros", version = "0.1.0", registry = "ghp-cargo-local" }
quanta = "0.11.1"
once_cell = "1.18.0"
chrono = { version = "0.4.26", default-features = false, features = ["clock"] }
rtrb = "0.2.3"
dyn-fmt = "0.4.0"

[dev-dependencies]
criterion = "0.4.0"
delog = "0.1.6"
tracing = "0.1.37"
tracing-appender = "0.2.2"
tracing-subscriber = "0.3.16"
recycle-box = { version = "0.2.0"}
trybuild = "1.0.85"

[[bench]]
name = "logger_benchmark"
harness = false

[[bench]]
name = "quicklog_benchmark"
harness = false

[badges]
maintenance = { status = "actively-developed" }
