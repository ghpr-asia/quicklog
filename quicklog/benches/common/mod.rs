use quicklog::serialize::Serialize;

#[macro_export]
macro_rules! loop_with_cleanup {
    ($bencher:expr, $loop_f:expr) => {
        loop_with_cleanup!($bencher, $loop_f, quicklog::logger().flush_noop())
    };

    ($bencher:expr, $loop_f:expr, $cleanup_f:expr) => {{
        quicklog::init!();

        $bencher.iter_custom(|iters| {
            let start = std::time::Instant::now();

            for _i in 0..iters {
                $loop_f;
                _ = $cleanup_f;
            }

            let end = start.elapsed();

            end
        })
    }};
}

#[allow(dead_code)]
#[derive(Clone, Copy, Debug)]
pub(crate) struct BigStruct {
    pub(crate) vec: [i32; 100],
    pub(crate) some: &'static str,
}

impl Serialize for BigStruct {
    fn encode<'buf>(&self, write_buf: &'buf mut [u8]) -> &'buf mut [u8] {
        // BigStruct is Copy, so we can just memcpy the whole struct
        let buf_ptr = write_buf.as_mut_ptr();
        let bytes = any_as_bytes(self);
        let n = bytes.len();
        let remaining = write_buf.len() - n;

        unsafe {
            buf_ptr.copy_from_nonoverlapping(bytes.as_ptr(), n);
            std::slice::from_raw_parts_mut(buf_ptr.add(n), remaining)
        }
    }

    fn decode(buf: &[u8]) -> (String, &[u8]) {
        let (chunk, rest) = buf.split_at(std::mem::size_of::<Self>());
        let bs: &BigStruct =
            unsafe { &*std::mem::transmute::<_, *const BigStruct>(chunk.as_ptr()) };

        (format!("{:?}", bs), rest)
    }

    #[inline]
    fn buffer_size_required(&self) -> usize {
        std::mem::size_of::<Self>()
    }
}

#[derive(Clone, Debug)]
pub(crate) struct Nested {
    pub(crate) vec: Vec<BigStruct>,
}

impl Serialize for Nested {
    fn encode<'buf>(&self, write_buf: &'buf mut [u8]) -> &'buf mut [u8] {
        self.vec.encode(write_buf)
    }

    fn decode(read_buf: &[u8]) -> (String, &[u8]) {
        let (vec, rest) = <Vec<BigStruct> as Serialize>::decode(read_buf);

        (format!("Nested {{ vec: {:?} }}", vec), rest)
    }

    #[inline]
    fn buffer_size_required(&self) -> usize {
        self.vec.buffer_size_required()
    }
}

pub(crate) fn any_as_bytes<T: Sized>(a: &T) -> &[u8] {
    unsafe { std::slice::from_raw_parts(a as *const T as *const u8, std::mem::size_of::<T>()) }
}
