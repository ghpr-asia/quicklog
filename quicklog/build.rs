use std::env;
use std::fs::File;
use std::io::Write;
use std::str::FromStr;

fn parse_value_from_config_with_default<T: FromStr>(
    key: &str,
    default: Option<T>,
) -> Result<T, String> {
    // Retrieve the value of the environment variable
    match env::var(key) {
        Ok(value) => match value.parse::<T>() {
            Ok(val) => Ok(val),
            Err(_) => {
                Err(format!("env var '{}' with value '{}' cannot be parsed into type '{2}'. Please set an env var can be parsed into '{2}'", key, value, std::any::type_name::<T>()))
            }
        },
        Err(_) => match default {
            Some(val) => Ok(val),
            None => {
                eprintln!("MAX_LOGGER_CAPACITY environment variable is not set");
                Err(format!("env '{}' is not set and there are no defaults for it. Please set it in your env.", key))
            }
        },
    }
}

fn main() {
    println!("cargo:rerun-if-env-changed=QUICKLOG_MAX_SERIALIZE_BUFFER_CAPACITY");
    let max_buffer_capacity = match parse_value_from_config_with_default(
        "QUICKLOG_MAX_SERIALIZE_BUFFER_CAPACITY",
        Some(1_000_000_usize),
    ) {
        Ok(val) => val,
        Err(err) => {
            println!("cargo:warning={}", err);
            1_000_000
        }
    };

    println!("cargo:rerun-if-env-changed=QUICKLOG_MAX_LOGGER_CAPACITY");
    let max_logger_capacity = match parse_value_from_config_with_default(
        "QUICKLOG_MAX_LOGGER_CAPACITY",
        Some(1_000_000_usize),
    ) {
        Ok(val) => val,
        Err(err) => {
            println!("cargo:warning={}", err);
            1_000_000
        }
    };

    // Generate the Rust source code
    let rust_code = format!(
        "// This file was generated by `build.rs`, do not modify this file manually!

/// Sets max capacity of logging queue, can be set through env var `QUICKLOG_MAX_LOGGER_CAPACITY`.
pub const MAX_LOGGER_CAPACITY: usize = {};

/// Sets max capacity of byte buffer used for serialization with `^` prefix in logging,
/// can be set through `QUICKLOG_MAX_SERIALIZE_BUFFER_CAPACITY`. 
pub const MAX_SERIALIZE_BUFFER_CAPACITY: usize = {};
",
        max_logger_capacity, max_buffer_capacity
    );

    // Write the code to a file
    let dest_path = std::path::Path::new("").join("src/constants.rs");
    let mut file = File::create(dest_path).expect("Failed to create file");
    file.write_all(rust_code.as_bytes())
        .expect("Failed to write file");
}
