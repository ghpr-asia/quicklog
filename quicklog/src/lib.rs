//! An asynchronous single-threaded logger where formatting and I/O are deferred at callsite.
//!
//! # Overview
//!
//! `Quicklog` is provides a framework for logging where it allows for deferred
//! deferred formatting and deferred I/O of logging, which should in turn provide
//! more performant logging with low callsite latency.
//!
//! ## Deferred Formatting
//!
//! #### Why?
//!
//! Formatting a `struct` into a `String` requires the overhead of serialization.
//! Deferring the serialization of a struct can be avoided by cloning / copying
//! the struct at a point in time, and saving that onto a queue.
//!
//! Later at the flush site, this struct is serialized into a string when I/O is
//! going to be performed.
//!
//! ## Deferred I/O
//!
//! #### Why?
//!
//! Deferring the I/O of formatting would allow for low callsite latency and allow
//! a user to implement their own flush site, possibly on a separate thread
//!
//! # Usage
//!
//! `init!()` macro needs to be called to initialize the logger before we can
//! start logging, probably near the entry point of your application.
//!
//! ## Example Usage
//!
//! ```
//! # use std::thread;
//! # use quicklog::{init, flush, info};
//! fn main() {
//!     init!();
//!
//!     // log some stuff
//!     info!("hello world! {}", "some argument");
//!
//!     // flush on separate thread
//!     thread::spawn(|| {
//!         loop {
//!             flush!();
//!         }
//!     });
//! }
//! ```
//!
//! # Macros
//!
//! #### Shorthand Macros
//!
//! Quicklog allows a number of macros with 5 different levels of verbosity. They are:
//!
//! * [`trace!`]
//! * [`debug!`]
//! * [`info!`]
//! * [`warn!`]
//! * [`error!`]
//!
//! ## Setup Macros
//!
//! Quicklog allows a user specified [`Clock`] or [`Flush`] to be implemented by
//! the user. This can be passed in through these macros, as long as the
//! underlying struct implements the correct traits
//!
//! * [`with_clock!`]: Specify the Clock Quicklog uses
//! * [`with_flush!`]: Specify the Flusher Quicklog uses
//! * [`with_flush_into_file`]: Specify path to flush log lines into
//!
//! ## Macro prefix for partial serialization
//!
//! To speed things up, if you are logging a large struct, there could be some small things
//! you might not want to log. This functionality can be done through implementing the
//! [`Serialize`] trait, where you can implement how to copy which parts of the struct.
//!
//! This could additionally be helpful if you already have the struct inside a buffer in byte
//! form, as you could simply pass the buffer directly into the decode fn, eliminiating any
//! need to copy.
//!
//! ```ignore
//! # use quicklog::{init, info, serialize::{Serialize, Store}};
//! struct SomeStruct {
//!     num: i64
//! }
//!
//! impl Serialize for SomeStruct {
//!    fn encode<'buf>(&self, write_buf: &'buf mut [u8]) -> (Store<'buf>, &'buf mut[u8]) { /* some impl */ }
//!    fn decode(read_buf: &[u8]) -> (String, &[u8]) { /* some impl */ }
//!    fn buffer_size_required(&self) -> usize { /* some impl */ }
//! }
//!
//! fn main() {
//!     init!();
//!     let s = SomeStruct { num: 1_000_000 };
//!     info!(^s, "some struct:");
//! }
//! ```
//!
//! ## Macro prefix for eager evaluation
//!
//! There are two prefixes you can use for variables, `%` and `?`. This works the same
//! way as `tracing`, where `%` eagerly evaluates an object that implements `Display`
//! and `?` eagerly evaluates an object that implements `Debug`.
//!
//! ```
//! # use quicklog::{init, info};
//! # fn main() {
//! # let impl_debug = "";
//! # let impl_display = "";
//! # init!();
//! info!(%impl_display, ?impl_debug);
//!
//! // logically expands into:
//! // info!(format!("impl_display={}", impl_display), format!("impl_debug={:?}", impl_debug));
//! # }
//! ```
//!
//! ## Structured fields
//!
//! Structured fields in log lines can be specified using `field_name = field_value`
//! syntax. `field_name` can be a literal or a bunch of idents. This can also
//! be used in combination with `%` and `?` prefix on args to eagerly evaluate
//! expressions into format strings.
//!
//! ```
//! # use quicklog::{init, info};
//! # fn main() {
//! # init!();
//! # let value = 10;
//! info!(question.answer = ?value, question.tricky=true, question.val= value, "some questions:");
//! // output: "some questions: question.tricky=true question.val=10 question.answer=10"
//! # }
//! ```
//!
//! # Environment variables
//!
//! There are two environment variables you can set:
//!
//! 1. `QUICKLOG_MAX_LOGGER_CAPACITY`
//!     - sets the size of the spsc ring buffer used for logging
//! 2. `QUICKLOG_MAX_SERIALIZE_BUFFER_CAPACITY`
//!     - sets the size of the byte buffer used for static serialization
//!     - this can be increased when you run into issues out of memory in debug
//!     when conducting load testing
//!
//! # Components
//!
//! ## quicklog-clock
//!
//! [`Clock`] is the trait for a clock that can be used with [`Quicklog`]. Clocks can
//! be swapped out at runtime with a different implementation.
//!
//! This swap should be done at the init stage of your application to ensure
//! that timings are consistent.
//!
//! ### Example
//!
//! ```ignore
//! struct SomeClock;
//!
//! impl Clock for SomeClock { /* impl methods */ }
//!
//! fn main() {
//!     init!();
//!
//!     with_clock!(SomeClock::new());
//!
//!     // logger now uses SomeClock for timestamping
//!     info!("Hello, world!");
//!     flush!();
//! }
//! ```
//!
//! ## quicklog-flush
//!
//! [`Flush`] is the trait that defines how the log messages would be flushed.
//! These logs can be printed through using the pre-defined [`StdoutFlusher`] or
//! saved to a file through the pre-defined [`FileFlusher`] to a specified
//! location through the string passed in.
//!
//! ### Example
//!
//! ```
//! # use quicklog::{info, init, flush, with_flush};
//! # use quicklog_flush::stdout_flusher::StdoutFlusher;
//! fn main() {
//!     init!();
//!
//!     with_flush!(StdoutFlusher);
//!     info!("hello world!");
//!
//!     // uses the StdoutFlusher passed in for flushing
//!     flush!();
//! }
//! ```
//!
//! [`Serialize`]: serialize::Serialize
//! [`StdoutFlusher`]: quicklog_flush::stdout_flusher::StdoutFlusher
//! [`FileFlusher`]: quicklog_flush::file_flusher::FileFlusher

use dyn_fmt::AsStrFormatExt;
use once_cell::unsync::Lazy;
use quanta::Instant;
use queue::{
    receiver::Receiver, sender::Sender, CursorRef, FlushError, FlushResult, LogArgType, Metadata,
    ReadError,
};
use rtrb::{chunks::WriteChunkUninit, RingBuffer};
use serialize::buffer::ByteBuffer;

pub use ::lazy_format;
pub use std::{file, line, module_path};

use chrono::{DateTime, Utc};
use quicklog_clock::{quanta::QuantaClock, Clock};
use quicklog_flush::{file_flusher::FileFlusher, Flush};

/// contains logging levels and filters
pub mod level;
/// contains macros
pub mod macros;
/// contains trait for serialization and pre-generated impl for common types and buffer
pub mod serialize;

include!("constants.rs");
/// `constants.rs` is generated from `build.rs`, should not be modified manually
pub mod constants;

pub mod queue;
mod utils;

pub use quicklog_macros::{debug, error, info, trace, warn, Serialize};

use crate::{queue::LogHeader, serialize::DecodeFn};

/// Logger initialized to Quicklog
#[doc(hidden)]
static mut LOGGER: Lazy<Quicklog> = Lazy::new(Quicklog::default);

/// **Internal API**
///
/// Returns a mut reference to the globally static logger [`LOGGER`]
#[doc(hidden)]
pub fn logger() -> &'static mut Quicklog {
    unsafe { &mut LOGGER }
}

pub trait PatternFormatter {
    fn custom_format(
        &mut self,
        time: DateTime<Utc>,
        metadata: &Metadata,
        log_record: &str,
    ) -> String;
}

pub struct QuickLogFormatter;

impl QuickLogFormatter {
    fn new() -> Self {
        Self {}
    }
}

impl PatternFormatter for QuickLogFormatter {
    fn custom_format(&mut self, time: DateTime<Utc>, _: &Metadata, log_record: &str) -> String {
        format!("[{:?}]{}\n", time, log_record)
    }
}

/// Main logging handler
pub struct Quicklog {
    flusher: Box<dyn Flush>,
    clock: Box<dyn Clock>,
    formatter: Box<dyn PatternFormatter>,
    sender: Sender,
    receiver: Receiver,
    fmt_buffer: String,
    byte_buffer: ByteBuffer,
}

impl Quicklog {
    /// Eagerly initializes the global [`Quicklog`] logger.
    /// Can be called through [`init!`] macro
    pub fn init() {
        // Referencing forces evaluation of Lazy
        _ = logger();
    }

    /// Sets which flusher to be used, used in [`with_flush!`]
    #[doc(hidden)]
    pub fn use_flush(&mut self, flush: Box<dyn Flush>) {
        self.flusher = flush
    }

    pub fn use_formatter(&mut self, formatter: Box<dyn PatternFormatter>) {
        self.formatter = formatter
    }

    /// Sets which clock to be used, used in [`with_clock!`]
    #[doc(hidden)]
    pub fn use_clock(&mut self, clock: Box<dyn Clock>) {
        self.clock = clock
    }

    /// Retrieves current [Instant](quanta::Instant).
    #[inline(always)]
    pub fn now(&self) -> Instant {
        self.clock.get_instant()
    }

    /// Internal API to get a chunk from buffer
    ///
    /// <strong>DANGER</strong>
    ///
    /// In release, the [`TAIL`] wraps around back to the start of the buffer when
    /// there isn't sufficient space left inside of [`BUFFER`]. If this happens,
    /// the buffer might overwrite previous data with anything.
    ///
    /// In debug, the method panics when we reach the end of the buffer
    #[doc(hidden)]
    pub fn get_chunk_as_mut(&mut self, chunk_size: usize) -> &mut [u8] {
        self.byte_buffer.get_chunk_as_mut(chunk_size)
    }

    /// Flushes all arguments from the queue.
    ///
    /// Iteratively reads through the queue to extract encoded logging
    /// arguments. This happens by:
    /// 1. Checking for a [`LogHeader`], which provides information about
    ///    the number of arguments to expect.
    /// 2. Parsing header-argument pairs.
    ///
    /// In the event of parsing failure, the flushing is terminated without
    /// committing.
    pub fn flush(&mut self) -> FlushResult {
        let chunk = self.receiver.read_chunk()?;
        let (head, tail) = chunk.as_slices();
        let mut cursor = CursorRef::new(head, tail);

        if cursor.is_empty() {
            return Err(FlushError::Empty);
        }

        loop {
            // Parse header for entire log message
            let Ok(log_header) = cursor.read::<LogHeader>() else {
                break;
            };

            let num_args = log_header.num_args;
            let mut decoded_args = Vec::with_capacity(num_args);
            while decoded_args.len() < num_args {
                let Ok(arg_type) = cursor.read::<LogArgType>() else {
                    break;
                };

                let decoded = match arg_type {
                    LogArgType::Fmt => {
                        // Remaining: size of argument
                        let size_of_arg = cursor.read::<usize>()?;
                        let arg_chunk = cursor.read_bytes(size_of_arg)?;

                        // Assuming that we wrote this using in-built std::fmt, so should be valid string
                        std::str::from_utf8(arg_chunk)
                            .map_err(|_| ReadError::UnexpectedValue)?
                            .to_string()
                    }
                    LogArgType::Serialize => {
                        // Remaining: size of argument, DecodeFn
                        let size_of_arg = cursor.read::<usize>()?;
                        let decode_fn = cursor.read::<DecodeFn>()?;
                        let arg_chunk = cursor.read_bytes(size_of_arg)?;

                        let (decoded, _) = decode_fn(arg_chunk);
                        decoded
                    }
                };
                decoded_args.push(decoded);
            }

            if decoded_args.len() != num_args {
                continue;
            }

            let time = self
                .clock
                .compute_system_time_from_instant(&log_header.instant)
                .expect("Unable to get time from Instant");
            let formatted = log_header.metadata.format_str.format(&decoded_args);
            let log_line = self
                .formatter
                .custom_format(time, log_header.metadata, &formatted);
            self.flusher.flush_one(log_line);
        }

        chunk.commit_all();

        Ok(())
    }

    /// Returns chunks/buffers for writing to
    pub fn prepare_write(&mut self) -> (WriteChunkUninit<'_, u8>, &mut String) {
        let chunk = self.sender.write_chunk().expect("queue full");
        let buf = &mut self.fmt_buffer;

        (chunk, buf)
    }

    /// Consumes a previously obtained write chunk and commits the written
    /// slots, making them available for reading
    pub fn finish_write(chunk: WriteChunkUninit<'_, u8>, committed: usize) {
        unsafe { chunk.commit(committed) }
    }
}

impl Default for Quicklog {
    fn default() -> Self {
        let (sender, receiver) = RingBuffer::<u8>::new(MAX_LOGGER_CAPACITY);

        Quicklog {
            flusher: Box::new(FileFlusher::new("logs/quicklog.log")),
            clock: Box::new(QuantaClock::new()),
            formatter: Box::new(QuickLogFormatter::new()),
            sender: Sender(sender),
            receiver: Receiver(receiver),
            fmt_buffer: String::with_capacity(2048),
            byte_buffer: ByteBuffer::new(),
        }
    }
}
