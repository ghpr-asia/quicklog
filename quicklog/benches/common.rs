use quicklog::serialize::{Serialize, Store};

#[macro_export]
macro_rules! loop_with_cleanup {
    ($bencher:expr, $loop_f:expr) => {
        loop_with_cleanup!($bencher, $loop_f, { quicklog::flush!() })
    };

    ($bencher:expr, $loop_f:expr, $cleanup_f:expr) => {{
        quicklog::init!();

        $bencher.iter_custom(|iters| {
            let start = quicklog::Quicklog::now();

            for _i in 0..iters {
                $loop_f;
            }

            let end = start.elapsed();

            $cleanup_f;

            end
        })
    }};
}

#[allow(dead_code)]
#[derive(Clone, Copy, Debug)]
pub(crate) struct BigStruct {
    pub(crate) vec: [i32; 100],
    pub(crate) some: &'static str,
}

impl Serialize for BigStruct {
    fn encode<'buf>(&self, write_buf: &'buf mut [u8]) -> (Store<'buf>, &'buf mut [u8]) {
        // BigStruct is Copy, so we can just memcpy the whole struct
        let (chunk, rest) = write_buf.split_at_mut(self.buffer_size_required());
        chunk.copy_from_slice(any_as_bytes(self));

        (Store::new(Self::decode, chunk), rest)
    }

    fn decode(buf: &[u8]) -> (String, &[u8]) {
        let (chunk, rest) = buf.split_at(std::mem::size_of::<Self>());
        let bs: &BigStruct =
            unsafe { &*std::mem::transmute::<_, *const BigStruct>(chunk.as_ptr()) };

        (format!("{:?}", bs), rest)
    }

    fn buffer_size_required(&self) -> usize {
        std::mem::size_of::<Self>()
    }
}

#[derive(Clone, Debug)]
pub(crate) struct Nested {
    pub(crate) vec: Vec<BigStruct>,
}

impl Serialize for Nested {
    fn encode<'buf>(&self, write_buf: &'buf mut [u8]) -> (Store<'buf>, &'buf mut [u8]) {
        self.vec.encode(write_buf)
    }

    fn decode(read_buf: &[u8]) -> (String, &[u8]) {
        let (vec, rest) = <Vec<BigStruct> as Serialize>::decode(read_buf);

        (format!("Nested {{ vec: {:?} }}", vec), rest)
    }

    fn buffer_size_required(&self) -> usize {
        self.vec.buffer_size_required()
    }
}

pub(crate) fn any_as_bytes<T: Sized>(a: &T) -> &[u8] {
    unsafe { std::slice::from_raw_parts(a as *const T as *const u8, std::mem::size_of::<T>()) }
}
