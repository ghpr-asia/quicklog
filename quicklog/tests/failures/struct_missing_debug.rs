use quicklog::info;

#[path = "../common/mod.rs"]
mod common;

use common::SerializeStruct;

fn main() {
    let s1 = SerializeStruct {
        symbol: "Hello world 1".to_string(),
    };
    info!(?s1, "struct does not implement Debug");
}
