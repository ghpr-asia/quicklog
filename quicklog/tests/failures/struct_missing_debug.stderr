error[E0277]: `SerializeStruct` doesn't implement `std::fmt::Debug`
  --> tests/failures/struct_missing_debug.rs:12:5
   |
12 |     info!(?s1, "struct does not implement Debug");
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `SerializeStruct` cannot be formatted using `{:?}`
   |
   = help: the trait `std::fmt::Debug` is not implemented for `SerializeStruct`
   = note: add `#[derive(Debug)]` to `SerializeStruct` or manually `impl std::fmt::Debug for SerializeStruct`
   = note: this error originates in the macro `format_args` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider annotating `SerializeStruct` with `#[derive(Debug)]`
  --> tests/failures/../common/mod.rs
   |
   + #[derive(Debug)]
   |
